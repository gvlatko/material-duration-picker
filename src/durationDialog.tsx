import * as React from 'react'
import {ComponentType, useEffect, useState} from 'react'
import {Button, Dialog, DialogActions, DialogContent, DialogProps, Toolbar, Typography, useTheme} from "@mui/material";
import {DurationType, DurationView, Labels} from "./types";
import {DurationFieldsContainer, DurationFieldsContainerProps} from "./durationFieldsContainer";
import DefaultLabels from './defaultLabelsEn.json'
import {timeToDuration} from "./utils";

export type DurationDialogProps = DialogProps & {
  time: number | undefined;
  onDismiss: () => void;
  onAccept: (time: number | undefined) => void

  views: DurationView[]
  labels?: Labels
  formatDuration: (duration: DurationType) => string;

  DurationFieldsContainerProps?: Partial<DurationFieldsContainerProps>
  DurationFieldsContainerComp?: ComponentType<DurationFieldsContainerProps>;
}

export const DurationDialog = ({
  time: _time,
  onDismiss,
  onAccept,
  views,
  labels: _labels,
  formatDuration,
  DurationFieldsContainerProps,
  DurationFieldsContainerComp = DurationFieldsContainer,
  ...props
}: DurationDialogProps) => {
  const theme = useTheme()
  const [time, setTime] = useState<number | undefined>(_time)

  useEffect(() => {
    setTime(_time)
  }, [_time])

  const duration = timeToDuration(time)

  const labels = {
    ...DefaultLabels,
    ..._labels
  }

  return (
    <Dialog
      fullWidth
      onClose={() => onDismiss()}
      {...props}
    >
      <Toolbar style={{
        backgroundColor: theme.palette.primary.main,
        color: theme.palette.primary.contrastText
      }}>
        <Typography variant='h4'>
          {duration && formatDuration(duration)}
        </Typography>
      </Toolbar>
      <DialogContent>
        <DurationFieldsContainerComp views={views} value={time} setValue={setTime} {...DurationFieldsContainerProps}/>
      </DialogContent>
      <DialogActions>
        <Button
          onClick={() => {
            onDismiss()
          }}
          color='primary'
        >
          {labels.cancel}
        </Button>
        <Button
          onClick={() => {
            onAccept(time)
            onDismiss();
          }}
          color='primary'
        >
          {labels.ok}
        </Button>
      </DialogActions>
    </Dialog>
  );
}

