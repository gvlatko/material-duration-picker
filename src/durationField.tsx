import {duration, TextField, TextFieldProps} from "@mui/material";
import * as React from 'react'
import {ComponentType, useEffect, useState} from "react";
import {DurationType, DurationView, Labels} from "./types";
import DefaultLabels from "./defaultLabelsEn.json";

export type DurationFieldProps = TextFieldProps & {
  duration: DurationType,
  onConfirm: (value: number | undefined) => void;
  view: DurationView;
  labels?: Labels;
  TextFieldComp?: ComponentType<TextFieldProps>;
}

export const DurationField = ({
  duration,
  onConfirm,
  TextFieldComp = TextField,
  view,
  labels: _labels,
  ...props
}: DurationFieldProps) => {
  const labels = {
    ...DefaultLabels,
    ..._labels
  }

  const [value, setValue] = useState<number | undefined>(duration[view] )

  useEffect(() => {
    setValue(duration[view] )
  }, [duration, setValue])

  return (
    <TextFieldComp
      type="number"
      onChange={({target}) => {
        const num = +target.value
        setValue(!isFinite(num) || target.value === '' ? undefined : num);
      }}
      onBlur={() => {
        onConfirm(value)
      }}
      value={value ?? ''}
      fullWidth
      label={labels[view]}
      {...props}
    />
  )
}
