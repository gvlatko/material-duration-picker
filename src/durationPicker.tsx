import * as React from 'react'
import {ComponentType, useState} from "react";
import {TextField, TextFieldProps} from "@mui/material";
import {durationToTime, timeToDuration} from "./utils";
import {DurationDialog, DurationDialogProps} from "./durationDialog";
import {DurationType, DurationView} from "./types";


export type DurationPickerProps = Partial<TextFieldProps> & {
  value: number | undefined,
  onValueChange: (value: number | undefined) => void;
  formatDuration: (duration: DurationType) => string
  disableEditDialog?: boolean;
  views?: DurationView[]
  DurationDialogProps?: Partial<DurationDialogProps>
  DurationDialogComp?: ComponentType<DurationDialogProps>;

  label: string | undefined;


  TextFieldComp?: ComponentType<TextFieldProps>;
}

export const DurationPicker = ({
  DurationDialogProps,
  value,
  onValueChange,
  formatDuration,
  views = ['hours', 'minutes'],
  disableEditDialog,
    label = 'Duration',
  DurationDialogComp = DurationDialog,
  TextFieldComp = TextField,
  ...props
}: DurationPickerProps) => {
  const [open, setOpen] = useState(false)

  const duration = timeToDuration(value)

  return <>
    <TextFieldComp
      onClick={() => {
        !disableEditDialog && setOpen(true);
      }}
      value={formatDuration(duration)}

      {...props}
      label={label}
      InputLabelProps={{
        shrink: open || undefined,
        ...props.InputLabelProps
      }}
      inputProps={{
        readOnly: true,
        ...props?.inputProps
      }}
    />
    {!disableEditDialog && (
      <DurationDialogComp
        time={value}
        open={open}
        onDismiss={() => setOpen(false)}
        onAccept={(time) => {
          onValueChange(time)
        }}
        label={label}
        views={views}
        formatDuration={formatDuration}
        {...DurationDialogProps}
      />
    )}
  </>;
};
