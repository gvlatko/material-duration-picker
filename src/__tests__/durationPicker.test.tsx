import * as React from 'react'
import {render, waitFor} from "@testing-library/react";
import {DurationPicker} from "../durationPicker";
import * as DurationDialog from '../durationDialog'
import {DurationField} from "../durationField";
import {timeToDuration} from "../utils";


jest.spyOn(DurationDialog, 'DurationDialog')
  .mockImplementation(({time}: any)=> <div>{JSON.stringify(timeToDuration(time))}</div>)

describe('durationPicker', () => {
  const onValueChange = jest.fn()
  const formatDuration = jest.fn()
  //1 week + 2 days + 3 hours + 4 min + 5 secs
  const time = 604800 + 172800 + 10800 + 240 + 5;

  it('can match snapshot', async () => {
    await waitFor(() => {
      expect(
        render(
          <DurationPicker
            value={time}
            onValueChange={onValueChange}
            formatDuration={formatDuration}
            label="Duration"

          />
        ).asFragment()
      ).toMatchSnapshot();

    })
  })
})
