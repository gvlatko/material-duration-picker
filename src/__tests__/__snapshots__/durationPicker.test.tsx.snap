// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`durationPicker can match snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
  >
    <label
      class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="false"
    >
      Duration
    </label>
    <div
      class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <input
        aria-invalid="false"
        class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
        readonly=""
        type="text"
        value=""
      />
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-173wfxe"
        >
          <span>
            Duration
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <div>
    {"weeks":1,"days":2,"hours":3,"minutes":4,"seconds":5}
  </div>
</DocumentFragment>
`;
